File: CS/XRProgressBar.cs
using System.ComponentModel;
using System.Drawing;
using DevExpress.Utils.Serializing;
using DevExpress.XtraPrinting;
using DevExpress.XtraReports.UI;

namespace SmartTagCustomization {
    [ToolboxItem(true)]
    public class XRProgressBar : XRControl {
        private float progress = 0;
        private float maxVal = 100;

        [DefaultValue(100), XtraSerializableProperty]
        public float MaxValue {
            get { return this.maxVal; }
            set {
                if (value <= 0)
                    return;
                this.maxVal = value;
            }
        }

        [DefaultValue(0), Bindable(true), XtraSerializableProperty]
        public float Progress {
            get { return this.progress; }
            set {
                if (value < 0 || value > maxVal)
                    return;
                this.progress = value;
            }
        }
        [Browsable(false)]
        public override string Text {
            get { return base.Text; }
            set { base.Text = value; }
        }

        public XRProgressBar() {
            this.ForeColor = SystemColors.Highlight;
            this.Progress = 20.0f;
        }

        protected override VisualBrick CreateBrick(VisualBrick[] childrenBricks) {
            return new PanelBrick(this);
        }

        protected override void PutStateToBrick(VisualBrick brick, PrintingSystemBase ps) {
            base.PutStateToBrick(brick, ps);
            PanelBrick panel = (PanelBrick)brick;
            VisualBrick progressBar = new VisualBrick(this);
            progressBar.Sides = BorderSide.None;
            progressBar.BackColor = panel.Style.ForeColor;
            progressBar.Rect = new RectangleF(0, 0, panel.Rect.Width * (Progress / MaxValue),
                panel.Rect.Height);
            panel.Bricks.Add(progressBar);
        }
    }
}

File: VB/XRProgressBar.vb
Imports System.ComponentModel
Imports System.Drawing
Imports DevExpress.Utils.Serializing
Imports DevExpress.XtraPrinting
Imports DevExpress.XtraReports.UI

Namespace SmartTagCustomization

    <ToolboxItem(True)>
    Public Class XRProgressBar
        Inherits XRControl

        Private progressField As Single = 0

        Private maxVal As Single = 100

        <DefaultValue(100), XtraSerializableProperty>
        Public Property MaxValue As Single
            Get
                Return maxVal
            End Get

            Set(ByVal value As Single)
                If value <= 0 Then Return
                maxVal = value
            End Set
        End Property

        <DefaultValue(0), Bindable(True), XtraSerializableProperty>
        Public Property Progress As Single
            Get
                Return progressField
            End Get

            Set(ByVal value As Single)
                If value < 0 OrElse value > maxVal Then Return
                progressField = value
            End Set
        End Property

        <Browsable(False)>
        Public Overrides Property Text As String
            Get
                Return MyBase.Text
            End Get

            Set(ByVal value As String)
                MyBase.Text = value
            End Set
        End Property

        Public Sub New()
            ForeColor = SystemColors.Highlight
            Progress = 20.0F
        End Sub

        Protected Overrides Function CreateBrick(ByVal childrenBricks As VisualBrick()) As VisualBrick
            Return New PanelBrick(Me)
        End Function

        Protected Overrides Sub PutStateToBrick(ByVal brick As VisualBrick, ByVal ps As PrintingSystemBase)
            MyBase.PutStateToBrick(brick, ps)
            Dim panel As PanelBrick = CType(brick, PanelBrick)
            Dim progressBar As VisualBrick = New VisualBrick(Me)
            progressBar.Sides = BorderSide.None
            progressBar.BackColor = panel.Style.ForeColor
            progressBar.Rect = New RectangleF(0, 0, panel.Rect.Width * (Progress / MaxValue), panel.Rect.Height)
            panel.Bricks.Add(progressBar)
        End Sub
    End Class
End Namespace
